crates/rpc/rpc/src/eth/api/changes.rs

```diff
@@ -0,0 +1,76 @@
+use crate::eth::error::{EthApiError, EthResult};
+use crate::EthApi;
+use reth_primitives::{Account, Address, BlockId, U256};
+use reth_provider::{AccountExtReader, BlockReaderIdExt, StateProviderFactory};
+use revm_primitives::HashMap;
+
+impl<Provider, Pool, Network> EthApi<Provider, Pool, Network>
+where
+    Provider: AccountExtReader + BlockReaderIdExt + StateProviderFactory + 'static,
+{
+    pub(crate) fn account_changes_in_block(
+        &self,
+        block_id: impl Into<BlockId>,
+    ) -> EthResult<Option<HashMap<Address, Account>>> {
+        let block_id = block_id.into();
+        let Some(block_number) = self.provider().block_number_for_id(block_id)? else {
+            return Err(EthApiError::UnknownBlockNumber)
+        };
+
+        let changed_accounts =
+            self.provider().changed_accounts_with_range(block_number..=block_number)?;
+        let state = self.provider().state_by_block_id(block_id)?;
+
+        if changed_accounts.is_empty() {
+            return Ok(None);
+        }
+
+        return changed_accounts.iter().try_fold(Some(HashMap::new()), |hash_map, &address| {
+            let Some(account) = state.basic_account(address)? else {
+                return Err(EthApiError::InternalEthError)
+            };
+
+            let mut hash_map = hash_map.unwrap();
+            hash_map.insert(address, account);
+            Ok(Some(hash_map))
+        });
+    }
+
+    pub(crate) fn balances_changes_in_block(
+        &self,
+        block_id: impl Into<BlockId>,
+    ) -> EthResult<Option<HashMap<Address, U256>>> {
+        let block_id = block_id.into();
+        let Some(block_number) = self.provider().block_number_for_id(block_id)? else {
+            return Err(EthApiError::UnknownBlockNumber);
+        };
+
+        if block_number == 0 {
+            return Err(EthApiError::Unsupported("balance changes on genesis"));
+        }
+
+        let prev_state = self.provider().state_by_block_number_or_tag((block_number - 1).into())?;
+
+        return match self.account_changes_in_block(block_id) {
```

Comment:
Ah this is not the most performant way of doing this. It should be similar (not to say the same) as `account_changes_in_block` 

Okay, figured out what is missing, you should add Provider function that fetches changeset. 
table `AccountChangeSet` contains all old values of changed accounts. So when you get that you already have a list of changed account and then you use `self.provider().state_by_` that gives you state after block executed so (you can use same block number for it) to fetch new values of accounts..

Reply:
Oh hmm, I think we also need to handle account destruction as well...  So the return might need to be `-> Result<Vec<Option<AccountBeforeTx>>>` in the provider and `-> RpcResult<HashMap<Address, Option<U256>>` for `reth_getBalanceChangesInBlock`..

Reply:
actually does `AccountChangeSet` include destructions?  I'm not seeing an obvious tombstone value but I might be missing something.

Reply:
> 



> Awesome, yup that's where I was headed originally but got lost in the weeds, it wasn't until I got a bit further that I realized that `changed_accounts_with_range` doesn't return the actual `AccountChangeSet` just the address.
> 
👍🏻  

> Question: should I make a new provider trait like `AccountChangeReader` or add this functionality to `AccountExtReader`? I'm not sure yet on the tradeoffs between having too many traits vs too many methods inside the traits. My gut is a new `AccountChangeReader` trait since `AccountExtReader` doesn't currently expose `AccountChangeSet` but I'm not sure if that's even the right metric to consider tbh.
> 
I would put it inside `AccountExtReader` it is not the perfect fit but it is good enough.

> Also, I noticed that the existing `AccountExtReader` methods return `BTreeMap|BTreeSet`, I assume since these are used in hash calculations so need to be ordered, but I'm wondering if the new methods should do the same for consistency?

Yeah `BtreeMap/Set` is needed for ordering of some data, but for changeset you can return ordinary `Vec` and write comments that it is already sorted as it will be.

> Oh hmm, I think we also need to handle account destruction as well... So the return might need to be `-> Result<Vec<Option<AccountBeforeTx>>>` in the provider and `-> RpcResult<HashMap<Address, Option<U256>>` for `reth_getBalanceChangesInBlock`..

For selfdestruct, wdyt about just setting zero balance?

`AccountBeforeTx` (aka changeset) contains Option of account, this means that account was not existing, https://github.com/paradigmxyz/reth/blob/d1610f3df9c8649de405ea139df366aa4863e635/crates/storage/db/src/tables/models/accounts.rs#L24 

but for selfdestruct state provider would return `None` for account.

Reply:
> For selfdestruct, wdyt about just setting zero balance?

Hmm, not a fan of conflating 0 and non-existent: if someone is tracking flows, they might want to know the difference between an account going to 0 naturally (e.g. spending all their eth) w/ the eth being burnt otherwise the flows won't add up.

My understanding on `AccountBeforeTx.info = None` is that that means the account was newly created in this block.  

> but for selfdestruct state provider would return None for account

Yes it will, but since I'm only iterating on accounts returned by the provider I won't know that the account's state needs to be included unless destructed accounts are included as well.

BTW I'm getting close to having another iteration ready for review but I have to run some errands so won't get back to it until later today.

Reply:
> > For selfdestruct, wdyt about just setting zero balance?
> 
> Hmm, not a fan of conflating 0 and non-existent: if someone is tracking flows, they might want to know the difference between an account going to 0 naturally (e.g. spending all their eth) w/ the eth being burnt otherwise the flows won't add up.

For example, BALANCE opcode returns zero if an account is not existing. So it is kinda already an equivalent, and with the removal of selfdestruct's this will not even show up. Just my train of thought, let us return more info I am fine with `Option`. 

> 
> My understanding on `AccountBeforeTx.info = None` is that that means the account was newly created in this block.

Yes.

> 
> > but for selfdestruct state provider would return None for account
> 
> Yes it will, but since I'm only iterating on accounts returned by the provider I won't know that the account's state needs to be included unless destructed accounts are included as well.

In changeset you would have account before it was destroyed (it is not different from just changed account), and only when you fetch account after this block you would get new state of that account and would see if it is `None` (self-destructed) or not.

`self.provider().state_by_block_id(block_id)?;` will allow you to fetch new values of the accounts.

> BTW I'm getting close to having another iteration ready for review but I have to run some errands so won't get back to it until later today.

It is 21pm here, take your time we are all async here :D

Reply:
Thanks @rakita I've pushed my changes in [d33e7f3](https://github.com/paradigmxyz/reth/pull/3768/commits/d33e7f32c4fee4d0ef31f1cb3291b804f6bfc950) I'll definitely  fixup the commits later I just figured it might be easier to see the changes if I left them unsquashed.

In addition to your suggested changes I've moved everything over to the new `RethApi` as my understanding is that's the desired final home, and I didn't want to polute `EthApi` in the meantime.

I'm still using a separate `AccountChangeReader` trait vs adding to `AccountExtReader`, but either way I think I'm going to run into the issue that I mentioned on Discord, namely:

> I'm having trouble figuring out how to make BlockChainProvider implement my new provider trait.  I think this would also still be an issue if I had simply added methods to AccountExtReader as I don't see it implementing that trait either, but I'm having trouble following along with how the BlockChainProvider and ProviderFactory<DB> interact.

This isn't an issue with the code at the moment but will be as soon as I try to actually register the and consume the new `reth_` namespace.

Reply:
> Thanks @rakita I've pushed my changes in [d33e7f3](https://github.com/paradigmxyz/reth/pull/3768/commits/d33e7f32c4fee4d0ef31f1cb3291b804f6bfc950) I'll definitely fixup the commits later I just figured it might be easier to see the changes if I left them unsquashed.
> 
> In addition to your suggested changes I've moved everything over to the new `RethApi` as my understanding is that's the desired final home, and I didn't want to polute `EthApi` in the meantime.
> 
> I'm still using a separate `AccountChangeReader` trait vs adding to `AccountExtReader`, but either way I think I'm going to run into the issue that I mentioned on Discord, namely:
> 

Left a few nitpick comments, the main things are there. Nice work!

> > I'm having trouble figuring out how to make BlockChainProvider implement my new provider trait.  I think this would also still be an issue if I had simply added methods to AccountExtReader as I don't see it implementing that trait either, but I'm having trouble following along with how the BlockChainProvider and ProviderFactory interact.
> 
> This isn't an issue with the code at the moment but will be as soon as I try to actually register the and consume the new `reth_` namespace.

Matt responded on this I think

Reply:
Awesome, thanks, I'll take a look at your comments and ya Matt replied on Discord so I'll try to get that working as well, thanks again for your patience and guidance!  I'll squash/fixup as needed then mark the PR as ready for review once I think it's all good to go.

Reply:
Ok pretty sure everything in this thread has been resolved, and since it's rather long now anyways probably best to re-raise anything I missed in the final review.